### DISCUSSION TITLE
Reading stream objects.

### DISCUSSION THREAD

**Discussion by silicontrip** [qpdf/qpdf#502] [Q&A]
Hi Jay,
 I'm getting inconsistent results reading object streams.
I start with a QPDFObjectHandle. Confirm it is a stream with isStream(). Get a PointerHolder<Buffer>. Call getPointer() to get a buffer. Then call getSize() and getBuffer() to get the pointer and length.  Sometimes I get the stream data and sometimes I get garbage memory. 
It feels like I need to retain the handle object. Or the content is being overwritten, as it's not associated with the handle when I store the handle into my custom hash array.
Do you have some sample code showing how to get stream data?
I remember reading something about retain counts in the headers, but that interface is private, I'm assuming it's not for me to explicitly retain the objects?
Thanks
Mark

**Comment by jberkenbilt** [comment]
Sounds like you're just getting raw rather than filtered stream data. What method are you using to get the stream data? Are you just using getStreamData()? That should give you the filtered stream data or throw an exception, but it depends on what methods you're calling and how you're calling them. When you say garbage memory, what do you mean? I don't think qpdf ever returns garbage memory. However if you allow the PointerHolder to go out of scope, the memory it points to might be deleted, but that would probably result in a segmentation fault. In any case, I need a little more to go on. In the source distribution, there are plenty of example programs in the examples directory, though I don't remember off hand which ones might get stream data in this way. There are lots of ways to get stream data, filter it, etc.

**Reply by silicontrip** [reply]
I'm storing a pointer to the QPDFObjectHandle, in my own wrapper class, which is stored in a tree structure.   It's then pulled back out of the tree, as a pointer, when needed. Then this is the code to get the stream, once the `QPDFObjectHandle` has been obtained.
```
		if (qpdf->isStream()) {
			try {
				PointerHolder<Buffer> bufRef = qpdf->getStreamData();
				Buffer* buf = bufRef.getPointer();
				size_t sz = buf->getSize();
				unsigned char * bb = buf->getBuffer();
```
This is the code that obtains the `QPDFObjectHandle`, before wrapping and storing it.
```
	if (pdfitem->isArray())
	{
		QPDFObjectHandle* thisObject = new QPDFObjectHandle( pdfitem->getArrayItem((int)index));
...
	} else if (pdfitem->isDictionary()) {
		// NSLog(@"obj is dictionary. child:ofItem:");
		
		std::set<std::string> keys = pdfitem->getKeys();  // I need this to be order preserving.
		std::vector<std::string> ord(keys.begin(), keys.end());
		std::set<std::string>::iterator iterKey;
		int loopindex=0;
		for(iterKey = keys.begin(); iterKey != keys.end(); ++iterKey)
		{
			if (loopindex == index)
			{
				QPDFObjectHandle* thisObject = new QPDFObjectHandle(pdfitem->getKey(*iterKey));
```
I have 1 PDF file which does throw an exception,  so I assume I'm using the correct interface.
```
WARNING: NSData (offset 78): error decoding stream data for object 4 0: stream inflate: inflate: data: unknown compression method
```

By garbage memory, it looks to me that the buffer pointer is pointing to the wrong place.   I do get the exact same bytes each time that I display the stream during the same run, restarting the program results in different data but again this is the same each time the stream object is displayed.    I see that the Buffer pointer is pointing to different memory but has the same bytes.  So I suspect that the internal Buffer heap is being overwritten, and possibly not segfaulting because the memory is still owned by the program, but using it for something else.

This is a log of the buffer address changing
```
2021-02-01 19:27:11.538734+1100 pdftracer[58793:3416727] buffer size: 1860 addr: 11b9200
2021-02-01 19:28:36.822409+1100 pdftracer[58793:3416727] buffer size: 1860 addr: 2050000
```
I did notice that during this run the memory was overwritten and I ended up with these log messages;
```
2021-02-01 19:28:39.008515+1100 pdftracer[58793:3416727] buffer size: 6 addr: b434230
2021-02-01 19:28:41.316261+1100 pdftracer[58793:3416727] buffer size: 6 addr: 1413bce0
2021-02-01 19:28:43.895957+1100 pdftracer[58793:3416727] buffer size: 6 addr: b3e0f0
2021-02-01 19:28:49.324165+1100 pdftracer[58793:3416727] buffer size: 6 addr: e72e5b0
2021-02-01 19:28:56.316976+1100 pdftracer[58793:3416727] buffer size: 6 addr: d135a0
2021-02-01 19:28:58.151717+1100 pdftracer[58793:3416727] buffer size: 6 addr: e708460
2021-02-01 19:28:59.857447+1100 pdftracer[58793:3416727] buffer size: 6 addr: e630270
```
The buffer contained the text "stream" I'm not sure where that data originated from. 

Thanks for the help.

**Comment by jberkenbilt** [comment]
There are a few things here.

You don't have to dynamically allocate QPDFObjectHandles. They are small objects that already contain an embedded smart pointer. They are relatively cheap to copy, and they are just references to the underlying object, so, for example, in this code:

```c++
QPDFObjectHandle a = QPDFObjectHandle::newArray();
QPDFObjectHandle b = a;
b.appendItem(QPDFObjectHandle::newString("A"));
```

`a` and `b` both point to the same array, now with a single element. This is just FYI. I don't see that the cause if the problem, but you're going through more steps than necessary. I wonder how clear this is from the docs. It is a fundamental aspects of the design of the qpdf API, but maybe it isn't communicated clearly enough.

Regarding your buffer pointer, note that `PointerHolder<Buffer>` is semantically equivalent to `std::shared_ptr<Buffer>`. The `PointerHolder` class has been part of the qpdf API since the beginning and long predates any kind of smart pointer in the C++ standard library. For backward compatibility, we're stuck with it. But what it means is that the memory is owned by the PointerHolder instance. If that goes out of scope, you lose the memory. You can't pull the `Buffer*` out and expect it to remain valid after `bufRef` goes out of scope, and the same goes for the `unsigned char*` that you are pulling out of the buffer. You also can't transfer ownership of the pointers. You have only two choices: hang onto the `PointerHolder<Buffer>` or make a copy of the memory. A PointerHolder is also a lightweight object -- it's just a pointer and a reference count. You can copy them around all you want, and copies all point to the same underlying object.

Regarding your comment about needing getKeys() to be order preserving, I'm not sure what you mean. PDF dictionaries do not have the concept of key order, and `std::set<std::string>` is sorted, so it will always be sorted. Your code seems to be trying to pull out a specific dictionary item by index, which makes me wonder whether maybe a better approach is possible.

Anyway, seeing all this makes me suspect the most likely source of your problem is that you are letting the `PointerHolder<Buffer>` go out of scope and expecting the memory you pulled out of it to be valid. In modern C++, there are few reasons to work with regular pointers outside of very narrow scope. I pretty much only directly allocate memory with `new` if I'm immediately putting the memory into a smart pointer, and if I pull a real pointer out of a smart pointer, I typically do it just within some block scope or something so that you can tell at a glance that the pointer is valid.

**Reply by silicontrip** [reply]
One of the pre-built wrapper classes I was using required pointers, it could not handle pass by value as it treats foreign objects as opaque pointers. 
I did write my own custom wrapper which I was able to change to pass by value. My code passes QPDFObjectHandles around and when required pull the stream data out, which is copied into the UI Text View.   The buffer reading code and copy to the text view all happens in the same code block.  The very next line after this code is initialising a string with the buffer pointer and length. 

```
		if (qpdf.isStream()) {
			try {
				PointerHolder<Buffer> bufRef = qpdf.getStreamData();
				Buffer* buf = bufRef.getPointer();
				size_t sz = buf->getSize();
				unsigned char * bb = buf->getBuffer();
				
				NSLog(@"buffer size: %ld addr: %x",sz,bb);
				
				for (int i=0; i<sz;++i)
				{
					printf("%d ",*(bb+i));
				}
```
This code produces;
```
2021-02-02 01:52:55.146453+1100 pdftracer[62430:3643705] buffer size: 1860 addr: 416c000
49 0 0 0 0 0 0 40 181 108 133 255 127 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 40 181 108 133 255 127 0 0 21 49 0 0 0 0 0 0 64 9 210 132 255 127 0 0 55 0 0 0 0 0 0 0 248 182 101 133 255 127 0 0 0 0 0 0 0 ...
```
If that pattern of bytes looks familiar.  So should the above code produce valid stream data? 

The order preserving comment was for myself.  Order preserving is needed for my UI elements, as they request everything by index. I'm thinking i'll store the keys in an array, alongside the PDFDictionary object.

Thanks again, for the help.

**Comment by jberkenbilt** [comment]
That code looks fine. Appreciate the other explanations. Of course the pattern of bytes is data-dependent. For an image, any pattern of bytes is possible. Is this an image? Anyway, so far, that all looks fine. When I do testing on qpdf itself, I have a build that I run with clang with `-fsanitize=address -fsanitize=undefined` passed in. If you have memory errors, this will go a long way to helping you find them. Just pass these flags to all invocations of clang or clang++. You can use these with gcc/g++ as well, but I find clang is a little better at this than gcc.

**Reply by silicontrip** [reply]
I've attempted to convert all the QPDFObjectHandles and QPDF references to pass by value and I've broken all my code, it might take me a little while to work out where it is going wrong.

**Reply by silicontrip** [reply]
I'm even more suspicious, that the QPDF objects are not being retained correctly.  There doesn't appear to a mechanism for chaging the ref count.   I wrote a method to convert the current QPDF to a string.  When I include this method, the program works, but doesn't if I remove the string conversion.  This is all it does;
```
- (NSString*)pdfString
{
	QPDFWriter qpdfWriter(qDocument);
	qpdfWriter.setOutputMemory();
	qpdfWriter.write();
	
	Buffer* qBuf = qpdfWriter.getBuffer();
	unsigned char const* qBytes = qBuf->getBuffer();
	NSData* qPDFData = [NSData dataWithBytes:qBytes length:qBuf->getSize()];
		
	return [[NSString alloc] initWithData:qPDFData encoding:NSMacOSRomanStringEncoding];
}
```
Not sure which bit is helping the QPDF object stay in memory.

**Reply by silicontrip** [reply]
sorry for all the posts, debugging while writing these replies.  The qpdf  object stays in memory when I call the first 3 lines of the above method.  Does this make any sense to you?

**Comment by jberkenbilt** [comment]
Your code contains syntax I'm not familiar with. What are the square brackets in `return [[NSString alloc] initWithData:qPDFData encoding:NSMacOSRomanStringEncoding];`? That doesn't look like regular C++ to me. Also the `@` in NSLog. Are you coding in objective C? I have no experience with that and have no idea how it interacts with qpdf. There is probably something about the way memory is being handled in whatever you're using that neither one of us understands.

**Reply by silicontrip** [reply]
It is Objective-C++, but it is designed to compile native C++ code.   It does sound like I'm not correctly following the memory management rules and to be honest my understanding of memory management doesn't go much beyond C.  So I need to be asking questions like;  Does `QPDF::processMemoryFile` make a copy of the memory or relies on the buffer being retained?   And should the QPDF->m->data pointer always be pointing to the same memory address, when i inspect it in my debugger?

**Comment by jberkenbilt** [comment]
QPDF objects aren't copiable. For a given QPDF object, m->data should always point to the same value. For multiple QPDFObjectHandles that are copies of each other, all their obj->data should point to the same thing. This is actually true of all QPDFObjectHandles that point to the same thing even if they were not created as copies of each other. Before a QPDFObjectHandle is resolved, obj->data may be nullptr. They get resolved automatically when you do anything that requires them to know what kind of object they are. So if you do `QPDFObjectHandle a = dict.getKey("/X");`, at that moment, `a.obj->data` is `nullptr`. However if you then do `a.isString()` (for example), `a.obj->data` will point to something. If you then do the same thing with another object handle, its obj->data will point to the same place.

When you call `processMemoryFile`, qpdf does not make a copy of the memory. The buffer must be retained. I hadn't thought about that, but if your QPDF object was processed in this way and that memory disappeared, it could definitely explain what you're seeing.

**Reply by silicontrip** [reply]
I think it may have something to do with the lazy resolving, because I'm getting this warning at runtime;
```
WARNING: object 7 0: operation for dictionary attempted on object of type null: returning false for a key containment request
WARNING: object 7 0: operation for dictionary attempted on object of type null: returning null for attempted key retrieval
WARNING: memory (page tree node, offset 3856): /Type key should be /Page but is not; overriding
WARNING: object 7 0: operation for dictionary attempted on object of type null: ignoring key replacement request
```
And this is the PDF which is returned
```
%PDF-1.3
%ø˜¢˛
1 0 obj
<< /Type /Catalog >>
endobj
xref
0 2
0000000000 65535 f 
0000000015 00000 n 
trailer << /Root 1 0 R /Size 2 /ID [<b8c25adc02619c1fdf4242f467fb1f05><236f3282c4b22e4113cce0982f29d069>] >>
startxref
51
%%EOF
```
Which doesn't look like the empty PDF from QPDF.cc
But if I put this code immediately after the qDocument.processMemoryFile() everything works;
```
QPDFWriter qpdfWriter(qDocument);
qpdfWriter.setOutputMemory();
qpdfWriter.write();
```
I guess this work around is suitable.
Does this make any sense?

**Comment by jberkenbilt** [comment]
It doesn't really make sense to me. Lots of people use qpdf for lots of things in lots of places, and I've never seen anything like this. But if you've found a workaround that works for you, I suppose there's no harm in using it as long as we remember that we never found a root cause for the odd behavior.

**Comment by silicontrip** [comment]
Hi Jay, thanks for helping out, made me have a close look at my code and I decided to refactor it so that only 2 of my Classes contained C++ code, also I found a design pattern that allowed the rest of my application to be completely seperate from the C++.  After I got it to compile it worked first attempt, no need to write QPDF to memory after initialisation.  I had not been able to understand the way that memory ownership needed to be handed off to the correct class, mixing obj-c and c++ made it more of a nightmare.  If you're curious screenshot of the ui is here; https://github.com/silicontrip/pdftracer/blob/main/example2.png
Thanks again and I'm really looking forward to the new features.

**Comment by jberkenbilt** [comment]
This sounds great. Thanks for the follow-up and for asking your questions here in public where others can benefit from them.

Your PDF Tracer project sounds interesting. As you work on, please feel free to engage with me on ideas about how qpdf can make your work easier. I'm always interested in feedback on the API, especially if you ever find that qpdf is forcing you to do things in a convoluted way or reinvent the wheel. For example, I'm working on some code now that requires me to generate text elements and know the coordinates of the text that was just placed. While qpdf is not a rendering engine, it knows how to do a lot of things that would be necessary for this. There is a private QPDFMatrix class that can be used to manipulate transformation matrices. I decided to make the class part of the public API for 10.2 so that I can use it in my code. Basically I'm writing a bunch of code outside of the qpdf library initially, and then as I get the code working, I will go back through it and, whenever I think, "It would be better if qpdf would assist here," I can update the API and move stuff in. So if you find yourself wrapping code around qpdf or implementing functionality that seems like it is missing, let me know. I sometimes accept contributions of code, though I'm quite stingy about that. :-) More often, I accept ideas or code snippets that I expand into qpdf code.

Anyway, whether you choose to engage in this way or not, best of luck on your project.

**Reply by silicontrip** [reply]
My wrappers are pretty thin.  The only functionality I needed to add for my UI in `QPDFObjectHandle` was `isExpandable; (isArray || isDictionary)` and `childrenContainIndirect`, which is a bit more complicated, the reasoning for this is that I cannot `parse` a string with an indirect reference, even if I didn't modify that reference, so this was a flag to remove the isEditable property from the TextView.  So if I were to ask for a feature it would some mechanism to handle indirect objects in `parse`  I haven't begun the code to add objects to a PDF yet, so I'll base my interface for it on what the library allows, usually if an API makes sense then the UI also makes sense. If not I'm sure to say, Hay! would doing it this way make more sense?

**Comment by jberkenbilt** [comment]
Coincidentally, I was actually thinking over the last few days that allowing indirect object references in QPDFObjectHandle::parse might be a nice enhancement. It would enable you to write code like

```c++
    QPDFObjectHandle font = doSomethingToGetFont();
    auto resources = QPDFObjectHandle::parse(context, "<< /Font << /F1 " + font.unparse() + " >> >>");
```

instead of the somewhat more cumbersome (albeit slightly more efficient)

```c++
   QPDFObjectHandle font = doSomethingToGetFont();
   auto resources = QPDFObjectHandle::parse("<< /Font << >> >>");
   resources.getKey("/Font").replaceKey("/F1", font);
```

This might be an easy change. The `context` I passed above would be a `QPDF*`, which is needed to resolve indirect object references, so this would be a new API call, and no guarantees that's what it will look like. I will consider this for 10.2. I might use it in my current project. Otherwise, what you're saying makes sense.

**Reply by silicontrip** [reply]
Thanks for that code snippet.  I get most of my inspiration from reading the APIs, so seeing how to implement something that isn't in the API is quite useful.

**Reply by jberkenbilt** [reply]
qpdf 10.2 will include an example of using the new parse. Check the release notes when it comes out. I'm hoping for next week. But you'll be able to do something like `QPDFObjectHandle::parse(&q, "<< /Font << /F1 " + font.unparse() + " >> >>");`

**Comment by jberkenbilt** [comment]
10.2 will have `QPDFObjectHandle::parse(QPDF* context, std::string const& object_str, std::string const& object_description = "")`. This was a trivial change.

**Comment by silicontrip** [comment]
Probably should close this, it was totally my error.

**Reply by jberkenbilt** [reply]
Error or not, the form of parse that allows indirect object literals is a nice edition to the library that I would not have thought of without this question, so I'm glad you asked it.
